lib/phoenix_live_view/async_result.ex

```diff
@@ -0,0 +1,501 @@
+      end
+
+      def handle_async(:org, {:ok, fetched_org}, socket) do
+        %{org: org} = socket.assigns
+        {:noreply, assign(socket, :org, AsyncResult.ok(org, fetched_org))}
+      end
+
+      def handle_async(:org, {:exit, reason}, socket) do
+        %{org: org} = socket.assigns
+        {:noreply, assign(socket, :org, AsyncResult.exit(org, reason))}
+      end
+
+  `start_async/3` is used to fetch the organization asynchronously. The
+  `handle_async/3` callback is called when the task completes or exists,
+  with the results wrapped in either `{:ok, result}` or `{:exit, reason}`.
+  The `AsyncResult` module is used to direclty to update the state of the
+  async operation, but you can also assign any value directly to the socket
+  if you want to handle the state yourself.
+  '''
+  use Phoenix.Component
+
+  defstruct name: nil,
+            keys: [],
+            ok?: false,
+            state: :loading,
+            result: nil
+
+  alias Phoenix.LiveView.{AsyncResult, Socket}
+
+  @doc """
+  Defines a new async result.
+
+  By default, the state will be `:loading`.
+  """
+  def new(name) do
+    new(name, [name])
+  end
+
+  def new(name, keys) do
+    loading(%AsyncResult{name: name, keys: keys, result: nil, ok?: false})
+  end
+
+  @doc """
+  Updates the state of the result to `:loading`
+  """
+  def loading(%AsyncResult{} = result) do
+    %AsyncResult{result | state: :loading}
+  end
+
+  @doc """
+  Updates the state of the result to `{:error, {:canceled, reason}}`.
+  """
+  def canceled(%AsyncResult{} = result, reason) do
+    error(result, {:canceled, reason})
+  end
+
+  @doc """
+  Updates the state of the result to `{:error, reason}`.
+  """
+  def error(%AsyncResult{} = result, reason) do
+    %AsyncResult{result | state: {:error, reason}}
+  end
+
+  @doc """
+  Updates the state of the result to `{:exit, reason}`.
+  """
+  def exit(%AsyncResult{} = result, reason) do
+    %AsyncResult{result | state: {:exit, reason}}
+  end
+
+  @doc """
+  Updates the state of the result to `:ok` and sets the result.
+
+  The `:ok?` field will also be set to `true` to indicate this result has
+  completed successfully at least once, regardless of future state changes.
+  """
+  def ok(%AsyncResult{} = result, value) do
+    %AsyncResult{result | state: :ok, ok?: true, result: value}
+  end
+
+  @doc """
+  Renders an async assign with slots for the different loading states.
+
+  ## Examples
+
+  ```heex
+  <AsyncResult.with_state :let={org} assign={@org}>
+    <:loading>Loading organization...</:loading>
+    <:empty>You don't have an organization yet</:error>
+    <:error :let={{_kind, _reason}}>there was an error loading the organization</:error>
+    <:canceled :let={_reason}>loading canceled</:canceled>
+    <%= org.name %>
+  <AsyncResult.with_state>
+  ```
+  """
+  attr :assign, :any, required: true
+  slot :loading
+
+  # TODO decide if we want an canceled slot
+  slot :canceled
```

Comment:
Given cancelled is an error, I would say no, and let the user pattern match. In fact, I would not wrap on a cancelled tuple either. The user can wrap if they want.

I know we have been back and forth on this but what if we call `incomplete` the `exit` and `error` bit? Here is why: if a request completes and then fails, we still show the valid result. Therefore, the point is to show the failure only if it never has been `ok?`, which is what incomplete could mean?